### DISCUSSION TITLE
Persistence with actors and side effects

### DISCUSSION THREAD

**Discussion by danielkcz** [statelyai/xstate#1832] [Q&A]
Our app needs to do a few tasks on the boot before it can go any further. First of all, it should restore persisted data (if any). And of course, persist on change. The idea is to persist context data only, but I am not yet sure if it's sufficient and it's rather cumbersome, to be honest.

I've stumbled upon https://github.com/davidkpiano/xstate/discussions/1515 which recommends **event sourcing** as a possible solution of persistence. It sounds nice and easy. However, any event that would cause a non-deterministic result will be a problem.

One boot task is to validate if the auth token is present and valid. If not, **obtain a guest token** from the backend. I want to delegate this logic into a separate child machine and spawn it as a long-lived actor to be used for future auth purposes (login). This is where event sourcing fails as it would grab a new guest token every time.

Persisting and restoring context only kinda solves this problem, but that means any state that should normally live in actors would have to be replicated to the top-level machine for persistence purposes only. That does sound like a dangerous idea.

I've prepared a small example. It's not really working, data are not propagated from child to parent for some reason. It should be enough to demonstrate the idea though.

https://codesandbox.io/s/persistence-with-actors-and-side-effects-8uh36?file=/src/bootMachine.js

Chances are that my modeling skills are really bad and there is some nicer way to tackle this.

**Comment by Andarist** [comment]
I believe that persisting "systems" (a hierarchical group of actors/machines with a single root) is a challenge right now and that we should carefully design a solution for this problem.

Given the presented codesandbox your problem looks a little bit different though from what I have initially anticipated - could you confirm/deny that? From what I see in the codesandbox you are trying to restore stuff "manually" and you in fact don't want to snapshot (and later restore using `service.start(persistedSnapshot)`) the whole "system". Is there any particular reason behind this? Would rehydrating the whole "system" solve your issue in a more automated & general way?

**Reply by danielkcz** [reply]
@Andarist I suppose rehydrating would be a nice solution if it would work for actors. Or maybe I am doing something wrong, but serializing state to JSON only has an ID for actors, not their state.

https://codesandbox.io/s/persistence-with-actors-and-side-effects-forked-4dwxm?file=/src/index.js

Do I need to serialize each actor state separately? I am not sure how I would restore it afterward.

I found a related discussion with the conclusion that it's not possible to reliably do this now? https://github.com/davidkpiano/xstate/discussions/1683

**Reply by Andarist** [reply]
Yes, this is rather hard/impossible right now - but I believe that we should tackle this problem at some point.

**Reply by danielkcz** [reply]
That's unfortunate.

~~Do you at least see what's the problem with the initial example and why changes are not propagated to the parent (through `propagateChange` action in both actors)? For some reason, the `bootMachine` receives only `AUTH_VALID/LOCALE_ACTIVE` events, but not the `AUTH_CHANGE/LOCALE_CHANGE`. In both cases, it uses the same `sendParent` action. I've added more console logs there to show it's firing. I am rather clueless about this.~~

Nevermind, my bad. Forgot how actions work. This is obviously wrong...

```js
    actions: {
      propagateChange: (ctx) => {
        return actions.sendParent({
          type: "LOCALE_CHANGE",
          locale: ctx.locale,
        });
      },
    },
```

Should be like this...

```js
    actions: {
      propagateChange: actions.sendParent(ctx => ({
          type: "LOCALE_CHANGE",
          locale: ctx.locale,
      })),
    },
```

https://codesandbox.io/s/persistence-with-actors-and-side-effects-8uh36?file=/src/localeMachine.js

Sadly the `@xstate/inspect` (commented out in `index.js`) is not helpful, it doesn't show any events. Might be the fault of CodeSandbox, hard to say.

**Reply by tomByrer** [reply]
I'm not sure if one needs actors.

I built an Radio Group element where each Radio Button is it's own machine, & the Group is a 'master' machine.  When the Group loads, it gets an array of Button text which the Radio Button machine maker uses to loop over.  So I had to code my own 'spawning' function, and another to properly tear them down (preventing memory leaks I hope).
